File: libs/openssl/ems_wrapper/src/mac.cpp
#include "mac.hpp"

#include "errors.hpp"

using namespace mac;

cmac::cmac()
    : mac(EVP_MAC_fetch(nullptr, "CMAC", nullptr), &EVP_MAC_free), ctx(EVP_MAC_CTX_new(mac.get()), &EVP_MAC_CTX_free)
{
    if (!mac)
    {
        throw std::runtime_error("EVP_MAC_fetch failed");
    }
    if (!ctx)
    {
        throw std::runtime_error("EVP_MAC_CTX_new failed");
    }
}

auto cmac::create(const uint8_vector &key, const std::string &algorithm) -> std::unique_ptr<cmac>
{
    auto cmac_instance = std::unique_ptr<cmac>(new cmac());

    const auto params =
        std::array{OSSL_PARAM_construct_utf8_string("cipher",
                                                    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
                                                    const_cast<char *>(algorithm.c_str()), 0),
                   OSSL_PARAM_construct_end()};

    if (EVP_MAC_init(cmac_instance->ctx.get(), key.data(), key.size(), params.data()) != 1)
    {
        throw_openssl_error("EVP_MAC_init failed");
    }

    return cmac_instance;
}

void cmac::update(const uint8_vector &data) const
{
    if (EVP_MAC_update(ctx.get(), data.data(), data.size()) != 1)
    {
        throw_openssl_error("EVP_MAC_update failed");
    }
}

auto cmac::final() const -> uint8_vector
{
    size_t outl = 0;
    if (EVP_MAC_final(ctx.get(), nullptr, &outl, 0) != 1)
    {
        throw_openssl_error("EVP_MAC_final failed to get output length");
    }

    uint8_vector data(outl);
    if (EVP_MAC_final(ctx.get(), data.data(), &outl, outl) != 1)
    {
        throw_openssl_error("EVP_MAC_final failed to finalize");
    }

    return data;
}

File: core-modules/crypto/src/bindings/mac.rs
use std::{ffi::CString, ptr};
use crypto_openssl_sys::*;
use crate::bindings::api::{throw_openssl_error, OsslError};
use crate::ossl_check;

pub struct Cmac {
    mac: *mut EVP_MAC,
    ctx: *mut EVP_MAC_CTX,
}

impl Drop for Cmac {
    fn drop(&mut self) {
        unsafe {
            EVP_MAC_CTX_free(self.ctx);
            EVP_MAC_free(self.mac);
        }
    }
}

impl Cmac {
    /// Fetches CMAC and allocates a new context.
    pub fn new() -> Result<Self, OsslError> {
        let mac = unsafe { EVP_MAC_fetch(ptr::null_mut(), CString::new("CMAC").unwrap().as_ptr(), ptr::null_mut()) };
        if mac.is_null() {
            return Err(throw_openssl_error("EVP_MAC_fetch failed"));
        }
        let ctx = unsafe { EVP_MAC_CTX_new(mac) };
        if ctx.is_null() {
            unsafe { EVP_MAC_free(mac) };
            return Err(throw_openssl_error("EVP_MAC_CTX_new failed"));
        }
        Ok(Cmac { mac, ctx })
    }

    /// Creates and initializes a CMAC with the given key and cipher name.
    pub fn create(key: &[u8], cipher: &str) -> Result<Self, OsslError> {
        let mut cm = Self::new()?;

        let cname = CString::new("cipher").unwrap();
        let alg_c = CString::new(cipher).unwrap();
        let mut params: [OSSL_PARAM; 2] = unsafe {
            [
                OSSL_PARAM_construct_utf8_string(
                    cname.as_ptr(),
                    alg_c.as_ptr() as *mut _,
                    0,
                ),
                OSSL_PARAM_construct_end(),
            ]
        };

        ossl_check!(
            unsafe { EVP_MAC_init(cm.ctx, key.as_ptr(), key.len(), params.as_ptr()) },
            "EVP_MAC_init failed"
        );
        Ok(cm)
    }

    /// Feeds more data into the CMAC.
    pub fn update(&mut self, data: &[u8]) -> Result<(), OsslError> {
        ossl_check!(
            unsafe { EVP_MAC_update(self.ctx, data.as_ptr(), data.len()) },
            "EVP_MAC_update failed"
        );
        Ok(())
    }

    /// Finalizes and returns the MAC bytes.
    pub fn finalize(&mut self) -> Result<Vec<u8>, OsslError> {
        // first call to get output length
        let mut outl: usize = 0;
        ossl_check!(
            unsafe { EVP_MAC_final(self.ctx, ptr::null_mut(), &mut outl, 0) },
            "EVP_MAC_final failed to get output length"
        );

        let mut out = vec![0u8; outl];
        ossl_check!(
            unsafe { EVP_MAC_final(self.ctx, out.as_mut_ptr(), &mut outl, outl) },
            "EVP_MAC_final failed to finalize"
        );
        out.truncate(outl);
        Ok(out)
    }
}